independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,117 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicReference<JarFileReference> jarFileReference;
+
+    private final AtomicInteger readersCounter = new AtomicInteger(0);
+
+    private volatile boolean releasing = false;
+
+    private JarFileReference(AtomicReference<JarFileReference> jarFileReference, JarFile jarFile) {
+        this.jarFileReference = jarFileReference;
+        this.jarFile = jarFile;
+    }
+
+    /**
+     * Increase the readers counter of the jarFile and returns it.
+     *
+     * @return The jarFile wrapped by this reference or null if a release has been requested thus making the file no longer
+     *         accessible.
+     */
+    private JarFile readLockAcquire() {
+        readersCounter.incrementAndGet();
+        if (releasing) {
+            if (readersCounter.decrementAndGet() == 0) {
+                close();
+            }
+            return null;
+        }
+        return jarFile;
+    }
+
+    /**
+     * Decrease the readers counter of the jarFile.
+     * If the counter drops to 0 and a release has been requested also closes the jarFile.
+     */
+    private void readLockRelease() {
+        if (readersCounter.decrementAndGet() == 0 && releasing) {
+            close();
+        }
+    }
+
+    /**
+     * Ask to release this reference.
+     * If there are no readers currently accessing the jarFile also close it, otherwise defer the closing when the last reader
+     * will leave.
+     */
+    void release() {
+        releasing = true;
+        if (readersCounter.get() == 0) {
+            close();
```

Comment:
This can be avoided with https://github.com/quarkusio/quarkus/pull/40942/files#r1626283806: we can let who is the last to release to take care of actually closing, so asking for a close become a fire-and-forget thing, at worse.